### PR TITLE
alts: Perform full handshake in ALTS tests.

### PR DESCRIPTION
**Pull request by matthewstevenson88**
The ALTS tests in this repo do not perform full ALTS handshakes against a fake handshaker service. This leads to changes in this repo breaking e2e/interop tests once the change is imported internally. This PR will (hopefully) reduce the frequency with which that happens.

RELEASE NOTES: none

### CODE REVIEW

File: credentials/alts/internal/handshaker/service/service.go [hunk 1]
```diff
@@ -58,3 +58,21 @@ func Dial(hsAddress string) (*grpc.ClientConn, error) {
 	}
 	return hsConn, nil
 }
+
+// Close closes all open connections to the handshaker service.
+//
+// For testing purposes only.
```
Comment (cesarghali) [inline credentials/alts/internal/handshaker/service/service.go:64]: This seems like a generic close function. Why is it expected to be used in tests only?

**Review by cesarghali — APPROVED** [review]

File: credentials/alts/internal/handshaker/service/service.go (hunk 1 above)
Comment (matthewstevenson88) [reply credentials/alts/internal/handshaker/service/service.go:64]: In practice, the connection to the handshaker service is kept open until the application is killed, so this Close function is never needed/called.

However, in the e2e test, we need to do call this Close function so that we don't get a goroutine leak.

**Comment by matthewstevenson88** [comment]
@easwars Would you be able to review from the gRPC side?

There are no functional changes in this PR, just adding another test (to hopefully reduce the cycle of "merge PR, import PR, see failure, revert PR", as happened with #6077).

File: credentials/alts/alts_test.go [hunk 2]
```diff
@@ -22,12 +22,21 @@
...
 	"testing"
+	"time"
 
 	"github.com/golang/protobuf/proto"
+	grpc "google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/alts/internal/handshaker/service"
 	altspb "google.golang.org/grpc/credentials/alts/internal/proto/grpc_gcp"
+	"google.golang.org/grpc/credentials/alts/internal/testutil"
 	"google.golang.org/grpc/internal/grpctest"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
```
Comment (easwars) [inline credentials/alts/alts_test.go:39]: Please split this into two imports:

```
	testpb "google.golang.org/grpc/interop/grpc_testing"
	testgrpc "google.golang.org/grpc/interop/grpc_testing"
```

In OSS, generated code for messages are services are placed in the same package. But this is not the case in google3. So, unless we have separate imports for the two, there is no way we can rewrite them at import time to point to the appropriate ones in google3.

File: credentials/alts/alts_test.go [hunk 3]
```diff
@@ -287,6 +296,50 @@ func (s) TestCheckRPCVersions(t *testing.T) {
...
+		grpc.WithReturnConnectionError(),
+	}
+	conn, err := grpc.DialContext(ctx, serverAddress, dialOptions...)
+	if err != nil {
+		t.Fatalf("grpc.Dial(%v, %v) failed: %v", serverAddress, dialOptions, err)
+	}
+	defer conn.Close()
+
+	// Ping the server.
+	c := testpb.NewTestServiceClient(conn)
```
Comment (easwars) [inline credentials/alts/alts_test.go:332]: This would have to be `testgrpc.NewTestServiceClient`

File: credentials/alts/alts_test.go [hunk 4]
```diff
@@ -300,3 +353,49 @@ func versions(minMajor, minMinor, maxMajor, maxMinor uint32) *altspb.RpcProtocol
...
+
+func startServer(t *testing.T, handshakerServiceAddress string, wait *sync.WaitGroup) (stop func(), address string) {
+	lis, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Errorf("net.Listen(tcp, :0) failed: %v", err)
+	}
+	serverOpts := &ServerOptions{HandshakerServiceAddress: handshakerServiceAddress}
+	creds := NewServerCreds(serverOpts)
+	s := grpc.NewServer(grpc.Creds(creds))
+	testpb.RegisterTestServiceServer(s, &testServer{})
```
Comment (easwars) [inline credentials/alts/alts_test.go:382]: testgrpc

File: credentials/alts/alts_test.go [hunk 5]
```diff
@@ -300,3 +353,49 @@ func versions(minMajor, minMinor, maxMajor, maxMinor uint32) *altspb.RpcProtocol
...
+		defer wait.Done()
+		if err := s.Serve(lis); err != nil {
+			t.Errorf("s.Serve(%v) failed: %v", lis, err)
+		}
+	}()
+	return func() { s.Stop() }, lis.Addr().String()
+}
+
+type testServer struct {
+	testpb.UnimplementedTestServiceServer
```
Comment (easwars) [inline credentials/alts/alts_test.go:394]: testgrpc

File: credentials/alts/alts_test.go [hunk 6]
```diff
@@ -22,12 +22,21 @@
...
+	"context"
+	"net"
 	"reflect"
+	"runtime"
+	"sync"
 	"testing"
+	"time"
 
 	"github.com/golang/protobuf/proto"
+	grpc "google.golang.org/grpc"
```
Comment (easwars) [inline credentials/alts/alts_test.go:34]: Why is this import renamed?

File: credentials/alts/internal/testutil/testutil.go [hunk 7]
```diff
@@ -22,11 +22,14 @@ package testutil
...
 	"bytes"
 	"encoding/binary"
+	"fmt"
 	"io"
 	"net"
 	"sync"
 
+	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/credentials/alts/internal/conn"
+	altspb "google.golang.org/grpc/credentials/alts/internal/proto/grpc_gcp"
```
Comment (easwars) [inline credentials/alts/internal/testutil/testutil.go:32]: Same comment about splitting imports for messages and services.

File: credentials/alts/alts_test.go [hunk 8]
```diff
@@ -287,6 +296,50 @@ func (s) TestCheckRPCVersions(t *testing.T) {
...
+	defer wait.Wait()
+	stopHandshaker, handshakerAddress := startFakeHandshakerService(t, &wait)
+	defer stopHandshaker()
+	stopServer, serverAddress := startServer(t, handshakerAddress, &wait)
+	defer stopServer()
+
+	// Connect to the server, authenticating with ALTS and the fake handshaker service.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	vmOnGCP = true
```
Comment (easwars) [inline credentials/alts/alts_test.go:318]: I don't see where this is used. If this is a package global, then please set it at the top of the test, and reset it to its original value with a `defer`, and have a small comment about why it is being set etc.

File: credentials/alts/alts_test.go [hunk 9]
```diff
@@ -287,6 +296,50 @@ func (s) TestCheckRPCVersions(t *testing.T) {
...
+	creds := NewClientCreds(&ClientOptions{HandshakerServiceAddress: handshakerAddress})
+	dialOptions := []grpc.DialOption{
+		grpc.WithTransportCredentials(creds),
+		grpc.WithBlock(),
+		grpc.WithReturnConnectionError(),
```
Comment (easwars) [inline credentials/alts/alts_test.go:323]: Are these two dial options necessary? We recommend our users to not use these dial options. So, we shouldn't be using them in our tests unless there is a strong reason for the same.

And if we don't have a good reason to use these dial options, and we can get rid of them, then we can use `grpc.Dial` instead of `grpc.DialContext`.

File: credentials/alts/alts_test.go [hunk 10]
```diff
@@ -287,6 +296,50 @@ func (s) TestCheckRPCVersions(t *testing.T) {
 	}
 }
 
+func TestFullHandshake(t *testing.T) {
```
Comment (easwars) [inline credentials/alts/alts_test.go:299]: A comment explaining what the test is doing, and what it is verifying would be very helpful to new readers of the code.

File: credentials/alts/alts_test.go [hunk 11]
```diff
@@ -287,6 +296,50 @@ func (s) TestCheckRPCVersions(t *testing.T) {
...
+	}
+	conn, err := grpc.DialContext(ctx, serverAddress, dialOptions...)
+	if err != nil {
+		t.Fatalf("grpc.Dial(%v, %v) failed: %v", serverAddress, dialOptions, err)
+	}
+	defer conn.Close()
+
+	// Ping the server.
+	c := testpb.NewTestServiceClient(conn)
+	if _, err = c.UnaryCall(context.Background(), &testpb.SimpleRequest{}); err != nil {
```
Comment (easwars) [inline credentials/alts/alts_test.go:333]: Prefer using `grpc.Dial` above and use a context with a reasonable deadline here.

File: credentials/alts/alts_test.go [hunk 12]
```diff
@@ -300,3 +353,49 @@ func versions(minMajor, minMinor, maxMajor, maxMinor uint32) *altspb.RpcProtocol
 		MaxRpcVersion: version(maxMajor, maxMinor),
 	}
 }
+
+func startFakeHandshakerService(t *testing.T, wait *sync.WaitGroup) (stop func(), address string) {
+	listener, err := net.Listen("tcp", ":0")
```
Comment (easwars) [inline credentials/alts/alts_test.go:358]: We have a testutil to create a local TCP listener which also works well in forge. See: https://github.com/grpc/grpc-go/blob/master/internal/testutils/local_listener.go#L24

File: credentials/alts/alts_test.go [hunk 13]
```diff
@@ -300,3 +353,49 @@ func versions(minMajor, minMinor, maxMajor, maxMinor uint32) *altspb.RpcProtocol
...
+		}
+	}()
+	return func() { s.Stop() }, lis.Addr().String()
+}
+
+type testServer struct {
+	testpb.UnimplementedTestServiceServer
+}
+
+func (s *testServer) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
```
Comment (easwars) [inline credentials/alts/alts_test.go:397]: You could use our stubserver type here: https://github.com/grpc/grpc-go/blob/3489bb7d51741618f1b1e0c723ca6dc905118e3b/internal/stubserver/stubserver.go#L42, and thereby get rid of `startServer`. Would that be possible?

File: credentials/alts/internal/handshaker/service/service.go (hunk 1 above)
Comment (easwars) [reply credentials/alts/internal/handshaker/service/service.go:64]: Why can't you explicitly close the connection from the test?

File: credentials/alts/internal/testutil/testutil.go [hunk 14]
```diff
@@ -123,3 +126,154 @@ func MakeFrame(pl string) []byte {
 	copy(f[conn.MsgLenFieldSize:], []byte(pl))
 	return f
 }
+
+// FakeHandshaker is a fake implementation of the ALTS handshaker service.
+type FakeHandshaker struct {
+	altspb.HandshakerServiceServer
```
Comment (easwars) [inline credentials/alts/internal/testutil/testutil.go:132]: This should be `altsgrpc`

File: credentials/alts/internal/testutil/testutil.go [hunk 15]
```diff
@@ -123,3 +126,154 @@ func MakeFrame(pl string) []byte {
...
+}
+
+// DoHandshake performs a fake ALTS handshake.
+func (h *FakeHandshaker) DoHandshake(stream altspb.HandshakerService_DoHandshakeServer) error {
```
Comment (easwars) [inline credentials/alts/internal/testutil/testutil.go:136]: Frankly, it is a lot of code for a fake server? Is it at all possible to use a real server instead, and give the tests real credentials to go with. We use real TLS certs for most of our tests. I don't know how possible/feasible it is for ALTS though.

File: credentials/alts/alts_test.go (hunk 2 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go:41]: Done.

File: credentials/alts/alts_test.go (hunk 3 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Done.

File: credentials/alts/alts_test.go (hunk 4 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Done.

File: credentials/alts/alts_test.go (hunk 5 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Done.

File: credentials/alts/alts_test.go (hunk 6 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Done.

File: credentials/alts/internal/testutil/testutil.go (hunk 7 above)
Comment (matthewstevenson88) [reply credentials/alts/internal/testutil/testutil.go:33]: Done.

File: credentials/alts/alts_test.go (hunk 8 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go:324]: This is a global variable defined in [`alts.go`](https://github.com/grpc/grpc-go/blob/3489bb7d51741618f1b1e0c723ca6dc905118e3b/credentials/alts/alts.go#L58). I added a comment in the test explaining why we have to reset it to true.

File: credentials/alts/alts_test.go (hunk 9 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: SG, removed and switched to `grpc.Dial`.

File: credentials/alts/alts_test.go (hunk 10 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go:304]: Done.

File: credentials/alts/alts_test.go (hunk 11 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Done.

File: credentials/alts/alts_test.go (hunk 12 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Oh awesome! Thanks for pointing this out!

File: credentials/alts/alts_test.go (hunk 13 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Thanks for the pointer! I kept the `startServer` function (just to consolidate the logic of starting up the server), but used `StubServer` to implement it, which greatly simplifies the logic. :)

File: credentials/alts/internal/testutil/testutil.go (hunk 14 above)
Comment (matthewstevenson88) [reply credentials/alts/internal/testutil/testutil.go]: Done.

File: credentials/alts/internal/testutil/testutil.go (hunk 15 above)
Comment (matthewstevenson88) [reply credentials/alts/internal/testutil/testutil.go]: Ya I agree, it's an unfortunate situation. For some context: all of the tests that use real ALTS credentials are run in google3 once the code is imported, and (before this PR) there are no e2e-style tests for ALTS in the OSS repo. This was done for 2 reasons:

1. ALTS credentials are proprietary and we do not want them to be checked into an OSS repo.
2. Because of security policies, we are not allowed to generate "test ALTS private keys" and check them into an OSS repo.

Because of these restrictions, if we want to have some sort of e2e test for ALTS in the OSS repo (like is added in this PR), then we need to completely fake out the handshaker service. Totally agreed that this is not perfect, but it's the same situation as what is done in C-core (and it's on my TODO list to add this to Java as well).

File: credentials/alts/internal/handshaker/service/service.go (hunk 1 above)
Comment (matthewstevenson88) [reply credentials/alts/internal/handshaker/service/service.go:64]: Disclaimer: I'm not a Go expert so apologies if I'm missing something obvious. :)

The [`hsConnMap`](https://github.com/grpc/grpc-go/blob/3489bb7d51741618f1b1e0c723ca6dc905118e3b/credentials/alts/internal/handshaker/service/service.go#L36), which holds the connection, is not exported. So we have 2 options:

1. Add a test-only `Close` API to close any connections in the map, which we call in `TestFullHandshake`.
2. Call `Dial` to get the same connection and then manually close it. This feels hacky, because we never want users to call `Dial` (it should only be called by the internals of the ALTS libraries) and so I didn't want to set this precedent).

WDYT?

File: credentials/alts/alts_test.go (hunk 13 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go:403]: Sorry for all the commits: I tried doing this but it ended being super flaky, so reverted back to the original solution.

File: credentials/alts/alts_test.go [hunk 16]
```diff
@@ -300,3 +359,49 @@ func versions(minMajor, minMinor, maxMajor, maxMinor uint32) *altspb.RpcProtocol
...
 	}
 }
+
+func startFakeHandshakerService(t *testing.T, wait *sync.WaitGroup) (stop func(), address string) {
+	listener, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("LocalTCPListener() failed: %v", err)
+	}
+	s := grpc.NewServer()
+	altspb.RegisterHandshakerServiceServer(s, &testutil.FakeHandshaker{})
```
Comment (easwars) [inline credentials/alts/alts_test.go:369]: This should be `altsgrpc` with a separate import.

File: credentials/alts/internal/handshaker/service/service.go (hunk 1 above)
Comment (easwars) [reply credentials/alts/internal/handshaker/service/service.go:64]: Thanks for the exaplaination. How about calling the method `CloseForTesting` or something which makes is very obvious that the method is only for testing purposes. We do that regularly in our codebase. Thanks.

File: credentials/alts/internal/testutil/testutil.go (hunk 15 above)
Comment (easwars) [reply credentials/alts/internal/testutil/testutil.go]: Ack.

File: credentials/alts/alts_test.go (hunk 16 above)
Comment (matthewstevenson88) [reply credentials/alts/alts_test.go]: Done, sorry I missed this one. :)

File: credentials/alts/internal/handshaker/service/service.go (hunk 1 above)
Comment (matthewstevenson88) [reply credentials/alts/internal/handshaker/service/service.go:64]: Sure thing, done.

**Review by easwars — APPROVED** [review]

**Comment by easwars** [comment]
One of the tests is failing because of a flake which we recently fixed. I've kicked off another run of the tests. If it fails again, maybe rebase your branch to master.
